msprime/trees.py

```diff
@@ -159,6 +177,34 @@ def __str__(self):
 
 
 class EdgesetTable(_msprime.EdgesetTable):
+    """
+    Class for tables describing all edgesets in a tree sequence, of the form
+        left	right	parent	children
+        5.0	7.0	4	0,1
+        2.0	2.5	5	1,3
+        3.2	3.5	4	0,1
+        0.0	3.2	6	0,1,3
+    These describe the half-open genomic interval affected: `[left, right)`,
+    the `parent` and the `children` on that interval.
+
+    Requirements: to describe a valid tree sequence, a `EdgesetTable` (and
+    corresponding `NodeTable`, to provide birth times) must satisfy:
+        1. each list of children must be in sorted order,
+        2. any two edgesets that share a child must be nonoverlapping, and
+        3. the birth times of the `parent` in an edgeset must be strictly
+            greater than the birth times of the `children` in that edgeset.
+    Furthermore, for algorithmic requirements
+        4. the the table must be sorted by birth time of the `parent`, and
+        5. any two edgesets corresponding to the same `parent` must be nonoverlapping.
+    It is an additional requirement that the complete ancestry of each sample
+    must be specified, but this is harder to verify.
+
+    It is not required that all records corresponding to the same parent be
```

Comment:
Proposal (rationale below):
 - require that edgeset tables be sorted by parent ID and left endpoint to be loaded into a tree sequence
 - provide a method to sort edgeset tables
This avoids having two types of tree sequence, and only requires checking for proper order at the time the tables are loaded.

Rationale: thinking again, I wasn't actually wanting less strict tree sequences, just less strict edgeset tables.  So: should provide a method to sort tables (just edgeset tables:  "sorting" a node table requires re-indexing the edgeset table - no need to implement this unless we actually have a use case) and assume sortedness in tree sequences.

I set about to actually do this, and realized that doing it non-hackily requires either:
 - a way to reorder the .children array, or
 - a way to pull out specific edgesets from a table
I then set out to implement the latter, as it seemed like simple copying over of tree_sequence_get_edgeset and related things... but eventually discovered it was not so straightforward (e.g. a tree sequence's edgeset is not a edgeset_table_t ...).

So, here's a hacky implementation.  I'd put it in the code if it wasn't so terrible and didn't really belong in this pull request:
```
def sort_edgesets(edgesets, nodes):
    """
    Returns a copy of the EdgesetTable sorted by parent ID (and hence, by
    birth time of parent), and for each parent, sorted by left endpoint.
    """
    out = msprime.EdgesetTable()
    temp_ts = msprime.load_tables(nodes=nodes, edgesets=edgesets)
    for (u, l, k) in sorted(zip(edgesets.parent, edgesets.left, 
                                range(edgesets.num_rows))):
        # out.add_row(*edgesets.get_edgeset(k))
        out.add_row(*temp_ts._ll_tree_sequence.get_edgeset(k))
    return out
```
If we had a `get_edgeset(k)` method for edgeset tables, this could be an `EdgesetTable` method and wouldn't need the `nodes` argument.

This has the additional property that edgesets appear in the order the parents appear in the node table - so, in order by time + parent id?  (which is the same as by parent id)  I'm not sure if this would actually be useful for anything, since parent ID already gives the index, but it's something that can be checked internally from only the edgeset without having to consult the NodeTable, so maybe this should be the requirement.